File: javascript/refactor.js
let fizzBuzz = (number) => {
  var array = [];

    for (var i = 1; i <= number; i++) {
      array.push(i);

      if (i % 3 === 0) {
        array.splice(i - 1, 1, "Fizz");
      };

      if (i % 5 === 0) {
        array.splice(i - 1, 1, "Buzz");
      };

      if (i % 15 === 0) {
        array.splice(i -1, 1, "FizzBuzz");
      }
    }
  return array
}
// ------------------------------------------------------------------

let fizzBuzz = (number) => {
  var array = [];

    for (var i = 1; i <= number; i++) {

      if ( i % 3 === 0 && i % 5 === 0) {
        array.push("FizzBuzz"); }

      else if ( i % 3 === 0) {
        array.push("Fizz"); }

      else if ( i % 5 === 0) {
        array.push("Buzz"); }

      else {
        array.push(i);
        }
      }

    return array
    }

// -----------------------------------------------------------------------

let valueFor = (i) => {
 if ( i % 3 === 0 && i % 5 === 0) {
    return ("FizzBuzz"); }

  else if ( i % 3 === 0) {
    return ("Fizz"); }

  else if ( i % 5 === 0) {
    return ("Buzz"); }

  else {
    return (i);
    }
  }

let fizzBuzz = (number) => {
  var array = [];
  for (var i = 1; i <= number; i++) {
    array.push(valueFor(i))
  }
  return array
}

File: ruby/refactor.rb
# First Go

class FizzBuzz
  def self.call(length)
    a = Array.new(length) { |i| i+1 }
    b = a.map { |x| x % 3 === 0 && x % 5 === 0 ? 'FizzBuzz' : x }
    c = b.map { |x| x % 3 === 0 ? 'Fizz' : x }
    d = c.map { |x| x % 5 === 0 ? 'Buzz' : x }
    d
  end
end

# --------------------------------------------------------------------

class FizzBuzz
  def self.call(length)
    a = Array.new(length) { |i| i+1 }
    b = a.map { |x| x % 3 === 0 && x % 5 === 0 ? 'FizzBuzz' : x }
    c = b.map { |x| x % 3 === 0 ? 'Fizz' : x }
    d = c.map { |x| x % 5 === 0 ? 'Buzz' : x }
    d
    # implicit return, do not need D
  end

  # def replaceNumWithString(array) do (don't need this do)
  #   b = array.map { |x| x % 3 === 0 && x % 5 === 0 ? 'FizzBuzz' : x }
  #   c = b.map { |x| x % 3 === 0 ? 'Fizz' : x }
  #   d = c.map { |x| x % 5 === 0 ? 'Buzz' : x }
  #
  # end
end

# --------------------------------------------------------------------

class FizzBuzz
  def self.call(length)
    a = Array.new(length) { |i| i+1 }
    replaceNumWithString(a)
  end

  def self.replaceNumWithString(a)
    b = array.map { |x| x % 3 === 0 && x % 5 === 0 ? 'FizzBuzz' : x }
    c = b.map { |x| x % 3 === 0 ? 'Fizz' : x }
    c.map { |x| x % 5 === 0 ? 'Buzz' : x }
  end
end

# --------------------------------------------------------------------

class FizzBuzz
  def self.call(length)
    a = Array.new(length) { |i| i+1 }
    replaceNumWithString(a)
  end

  def self.replaceNumWithString(a)
    a.map { |x| x % 3 == 0 && x % 5 == 0 ? 'FizzBuzz' : x }
     .map { |x| x % 3 == 0 ? 'Fizz' : x }
     .map { |x| x % 5 == 0 ? 'Buzz' : x }
  end
end

# --------------------------------------------------------------------

class FizzBuzz
  def self.call(length)
    # a = Array.new(length) { |i| i+1 }
    replaceNumWithString(1.upto(length))
  end

  def self.replaceNumWithString(a)
    a.map { |x| x % 3 == 0 && x % 5 == 0 ? 'FizzBuzz' : x }
     .map { |x| x % 3 == 0 ? 'Fizz' : x }
     .map { |x| x % 5 == 0 ? 'Buzz' : x }
  end
end

# --------------------------------------------------------------------
# RANGE
class FizzBuzz
  def self.call(x, y)
    a = (x..y).to_a
    replaceNumWithString(a)
  end

  def self.replaceNumWithString(a)
    a.map { |x| x % 3 == 0 && x % 5 == 0 ? 'FizzBuzz' : x }
     .map { |x| x % 3 == 0 ? 'Fizz' : x }
     .map { |x| x % 5 == 0 ? 'Buzz' : x }
  end
end
